File: task-03/Elixir/Sub task 4/subtask4.exs
defmodule Diamond do
  def print_diamond(output_file, n) when n < 1 do
    IO.write(output_file, "Number should be greater than 1\n")
  end

  def print_diamond(output_file, n) do
    mid = div(n, 2)

    # Upper part of the diamond
    for i <- Enum.map(1..div(n, 2) + 1, &(&1 * 2 - 1)) do
      spaces = mid - div(i, 2)
      IO.write(output_file, String.duplicate(" ", spaces) <> String.duplicate("*", i) <> "\n")
    end

    # Lower part of the diamond
    lower_part =
      if rem(n, 2) == 0 do
        Enum.map(Enum.reverse(1..div(n, 2)), &(&1 * 2 - 1))
      else
        Enum.map(Enum.reverse(1..div(n - 1, 2)), &(&1 * 2 - 1))
      end

    for i <- lower_part do
      spaces = mid - div(i, 2)
      IO.write(output_file, String.duplicate(" ", spaces) <> String.duplicate("*", i) <> "\n")
    end
  end

  def main do
    case File.read("input.txt") do
      {:ok, content} ->
        n = String.trim(content) |> String.to_integer()

        File.open("output.txt", [:write], fn output_file ->
          print_diamond(output_file, n)
        end)

        IO.puts("File transfer was successful")

      {:error, reason} ->
        IO.puts("Failed to open input.txt: #{reason}")
    end
  end
end

Diamond.main()

File: task-03/Haskell/Sub task 4/subtask4.hs
import System.IO
import Control.Monad (when)

printDiamond :: Handle -> Int -> IO ()
printDiamond outputFile n
  | n < 1 = hPutStrLn outputFile "Number should be greater than 1"
  | otherwise = do
      let mid = n `div` 2

      -- Upper part of the diamond
      mapM_ (printLine outputFile mid) [1,3..n]

      -- Lower part of the diamond
      if even n
      then mapM_ (printLine outputFile (mid - 1)) (reverse [1,3..(n-1)])
      else mapM_ (printLine outputFile mid) (reverse [1,3..(n-2)])

printLine :: Handle -> Int -> Int -> IO ()
printLine outputFile mid i = do
  let spaces = mid - (i `div` 2)
  hPutStrLn outputFile (replicate spaces ' ' ++ replicate i '*')

main :: IO ()
main = do
  inputFile <- openFile "input.txt" ReadMode
  outputFile <- openFile "output.txt" WriteMode

  input <- hGetLine inputFile
  let n = read input :: Int

  printDiamond outputFile n

  hClose inputFile
  hClose outputFile
  
  putStrLn "File transfer was successful"

